src/Sparse/Wiedemann.jl

```diff
@@ -0,0 +1,235 @@
+	@info "wiedemann 3 used"
+    RR = base_ring(A)
+	N = modulus(RR)
+	#T= elem_type(RR)
+	#A = change_base_ring(RR, A)::SMat{T}
+	(n,m) = nrows(A),ncols(A);
+	# Prealloc +Randomchoice
+    r = rand(RR, m)
+	c = rand(RR, m)
+    randlin = rand_srow(min(m,10),m,min(10,N),RR)
+	seq = Vector{T}(undef, 2*m)
+	storing_n = zeros(RR,n)#Vector{T}(undef,n)
+    storing_m = zeros(RR,m)#Vector{T}(undef,m)
+	z = zero(RR)
+    #Wiedemann sequence
+    # solve A^tAx = y2 => x -y in kernel(A^tA) to avoid finding zero vec
+	y = multi!(storing_m,TA, multi!(storing_n,A,r))
+	seq[1] = dot(randlin,c,z) #randlin*c 		
+	for i = 2:2*m  #Wiedemann sequence
+        c =  multi!(c,TA, multi!(storing_n,A,c))
+		seq[i] = dot(randlin,c) 
+	end
+    ##########################################################################################################################################
+	done,f = Hecke.berlekamp_massey(seq)
+	delta =0
+	while iszero(coeff(f,0))
+		delta+=1
+		f = divexact(f,gen(parent(f)))
+	end
+	constpartoff = coeff(f,0)
+	a = -inv(constpartoff)
+	reducedf = divexact(f-constpartoff,gen(parent(f)))
+    #f(TA*A)'c
+    v = Hecke.evaluate(reducedf,TA,A,y).*a
+    return true, (v-r)
+end
+#(p-1)/2 probably not prime
+#=
+function wiedemann(A::SMat{T}, TA::SMat{T}) where T<:Union{nmod, fmpz_mod}
+	@info "wiedemann catch used"
+	RR = base_ring(A)
+	N = modulus(RR)
+	(n,m) = nrows(A),ncols(A);
+	# Prealloc +Randomchoice
+    r = rand(RR, m)
+	c = rand(RR, m)
+    randlin = rand_srow(min(m,10),m,min(10,N),RR)
+	seq = Vector{T}(undef, 2*n)
+	storing_n = zeros(RR,n)#Vector{T}(undef,n)
+    storing_m = zeros(RR,m)#Vector{T}(undef,m)
+    #Wiedemann sequence
+    # solve A^tAx = y2 => x -y in kernel(A^tA) to avoid finding zero vec
+	y = mul!(storing_m,TA, mul!(storing_n,A,r))
+	seq[1] = dot(randlin,c,RR(0)) #randlin*c 		
+	for i = 2:2*n  #Wiedemann sequence
+        c =  mul!(c,TA, mul!(storing_n,A,c))
+		seq[i] = dot(randlin,c) 
+	end
+	#=
+    coeff_ = try                     
+		Hecke.berlekamp_massey(seq)
+	catch
+		e = false	               #TODO: assign gcd here
+	end
+    if !coeff_
+		return false, 0
+	end
+	=#
+	delta =0
+	while iszero(coeff(f,0))
+		delta+=1
+		f = divexact(f,gen(parent(f)))
+	end
+	constpartoff = coeff(f,0)
+	a = -inv(constpartoff)
+	reducedf = divexact(f-constpartoff,gen(parent(f)))
+    #f(TA*A)'c
+    v = Hecke.evaluate(reducedf,TA,A,y).*a
+    return true, (v-r)
+end
+=# 
+function Hecke.evaluate(f,TA,A::SMat{T},c) where T <:Union{gfp_elem, nmod}
+    #return f(A^t *A)*c
+	#T = elem_type(base_ring(A))
+	(n,m) = size(A)
+	storing_n = Vector{T}(undef,n)
+    s =  Vector{T}(undef,m)
+	C = collect(coefficients(f))
+	n = length(C)
+	s =  Hecke.mul!(s,TA, Hecke.mul!(storing_n,A,c)).*C[end]+c.*C[end-1]
+	for i = n-2:-1:1
+		s =  Hecke.mul!(s,TA, Hecke.mul!(storing_n,A,s)) + c.*C[i]
+	end
+	return s
+end
+
+function Hecke.evaluate(f,TA,A::SMat{T},c) where T <:Union{gfp_fmpz_elem, fmpz_mod}
```

Comment:
Indentation. What is the difference to the function above?

Reply:
mul! and multi! again